A68-deterministic-subsetting-lb-policy.md

```diff
@@ -0,0 +1,177 @@
+    // excluded_count indicates how many leftover backends we have on every round.
+    excluded_count = backend_count % subset_size
+
+    // We want to choose what backends are excluded in a round robin fashion before shufling the backends.
+    // This way we make sure that for any 2 backends the difference of how many times they are excluded is at most 1.
+    // excluded_start indicates the index of the first excluded backend
+    excluded_start := (round * excluded_count) % backend_count
+    // excluded_start indicates the index after the last excluded backend.
+    // It could wrap around the end of the addresses list.
+    excluded_end := (excluded_start + excluded_count) % backend_count
+
+    if excluded_start < excluded_end {
+        // excluded_end doesn't wrap, exclude addresses from the interval [excluded_start, excluded_end)
+    } else {
+        // excluded_end wraps around the end of the addresses list, exclude intervals [0:excluded_start] and [excluded_end, end_of_the_array)
+    }
+
+    // randomly shuffle the addresses to increase subset diversity. Use round as seed to make sure
+    // that clients within the same round shuffle addresses identically. 
+    addresses.shuffle(seed: round)
+
+    // subset_id is the index for the current client withing the round
+    subset_id := client_index % subset_count
+
+    // calculate start start and end of the resulting subset 
+    start = subsetId * subset_size
+    end = start + subset_size
+    return addresses[start: end]
+}
+```
+
+
+### Characteristics of the selected algorithm
+
+* Every client connects to exactly `subset_size` backends.
+* Clients are evenly distributed between backends. The most connected backend receives at most 2 more connections than the least connected one. One comes from the fact that a backend can be excluded one more time than some other backend. And another one comes from the fact that there might be not enough client to completely fill the last round. 
+* Backends are randomly shuffled between clients. If a small subset of backends loses connectivity, the algorithm maximizes the chances that the backends from this subset will be evenly distributed between clients. 
+* The algorithm generates some, potentially significant, connection churn during backend scaling or redeployment. This might be a problem for very small subset sizes.
+
+### LB Policy Config and Parameters
+
+The `deterministic_subsetting` LB policy config will be as follows.
+
+```
+message LoadBalancingConfig {
+    oneof policy {
+        DeterministicSubsettingLbConfig deterministic_subsetting = 21 [json_name = "deterministic_subsetting"];
+    }
+}
+
+message DeterministicSubsettingLbConfig {
+    // client_index is an index within the
+    // interval [0..N-1], where N is the total number of clients.
+    // Every client must have a unique index.
+    google.protobuf.UInt32Value client_index = 1;
```

Comment:
This is really a continuation of the discussion about client_id, but I want to start a separate thread to keep it more focused.  What I am trying to do is to take a step back and think about options that don't involve client_id. 

One obvious option is a fully random subsetting, where every client selects N backends at random. The problem with this option is that resulting load on the backends end up being not even, which is described in more details [here](https://queue.acm.org/detail.cfm?id=3570937) But what if we can make it better by using WRR as a child balancer? My intuition is that WRR should make clients to send more requests to the least utilized backends within the subset, and if we have good subset diversity and relatively large subsets, WRR can, at least partially, compensate for backend load imbalance. 

If this is not enough we can extend this idea by creating a "load aware subset balancer" The balancer will do what I described above, but additionally it will periodically remove the most utilized backend from its subset, if the deviation in load for this backend diverged too far from the mean for the whole subset. This will probably require an additional ORCA metric, which reports pure CPU load on the backends. (I think the metric that is used by WRR is more `cost per request` which makes it not suitable to compare load on the backends)

@markdroth what do you think?

Reply:
> One obvious option is a fully random subsetting, where every client selects N backends at random. The problem with this option is that resulting load on the backends end up being not even, which is described in more details [here](https://queue.acm.org/detail.cfm?id=3570937) But what if we can make it better by using WRR as a child balancer? My intuition is that WRR should make clients to send more requests to the least utilized backends within the subset, and if we have good subset diversity and relatively large subsets, WRR can, at least partially, compensate for backend load imbalance.

WRR definitely does help balance load within the subset.  However, I think that solves the problem only for workloads where every client generates approximately the same load.  If the subsets are not evenly distributed and you happen to have a set of clients that generate more load than others and they all happen to be talking to the same set of backends, then you will still wind up with uneven load across all backends.


> If this is not enough we can extend this idea by creating a "load aware subset balancer" The balancer will do what I described above, but additionally it will periodically remove the most utilized backend from its subset, if the deviation in load for this backend diverged too far from the mean for the whole subset. This will probably require an additional ORCA metric, which reports pure CPU load on the backends. (I think the metric that is used by WRR is more `cost per request` which makes it not suitable to compare load on the backends)

If WRR is using the right weight for balancing load across the backends, then it's not clear to me that this kind of outlier detection-style backend ejection is actually necessary.  If one particular backend's weight is much lower than the others, WRR should give it much less traffic anyway, so ejecting the backend wouldn't really change anything.

Also, if the problem is that a particular set of clients is generating higher load than other clients, then ejecting some of the backends in the subset seems like it could result in even worse load imbalance, by forcing those heavy-traffic clients onto an even smaller set of backends.

Reply:
(To be clear, BTW, I am not necessarily objecting to implementing random subsetting -- there may very well be workloads for which that works fine.  I'm just pointing out other factors to consider here when evaluating whether a given algorithm will work for your particular workload.)

Reply:
> WRR definitely does help balance load within the subset. However, I think that solves the problem only for workloads where every client generates approximately the same load. If the subsets are not evenly distributed and you happen to have a set of clients that generate more load than others and they all happen to be talking to the same set of backends, then you will still wind up with uneven load across all backends.

I think this argument can be applied to any subsetting implementation, including the Google's one as well as Twitter's Aperture. All those algorithms don't take into account potential difference between clients and only balance number of connections per backend, not req per backend. I guess the way it works is by grouping clients into groups of identical instances. My intuition is that random subsetting should deal with this particular problem (different types of clients) just fine because the resulting subsets are fully random, which makes the probability of any group of clients talking to the same subset of backends almost neglectable. 

 > If WRR is using the right weight for balancing load across the backends, then it's not clear to me that this kind of outlier detection-style backend ejection is actually necessary. If one particular backend's weight is much lower than the others, WRR should give it much less traffic anyway, so ejecting the backend wouldn't really change anything.

> Also, if the problem is that a particular set of clients is generating higher load than other clients, then ejecting some of the backends in the subset seems like it could result in even worse load imbalance, by forcing those heavy-traffic clients onto an even smaller set of backends.

I think I described it poorly. When I said eject what I really meant is first eject the backend and then replace it with another random one, so we will always have exactly N backends per client. The problem with just WRR is that it can balance only requests and not connections per backend, and the proposed solution could also be used to balance the connections. For example, we can implement the load metric to take into account the number of incoming connections. instead of CPU load. Then top level subsetting LB will be balancing just the number of connections per backend and child LB (wrr) will be balancing load within the subset. I think that by tuning different types of load metrics we can achieve perfect load distribution across all backends, though I fully agree that it might not be necessary and just subsetting + wrr might be enough

What I am going to do is to test all those options on a synthetic benchmark. I'll get back to this PR when I have some results.

Reply:
I got some results testing random subsetting with WRR

**Test setup**
* 87 backends, each have 1 CPU but some backends are more powerful then the others
* 93 clients. Clients [0-69] are "normal" (they sleep between requests for 15 millisecond) The rest of the clients are "noizy"(they keep sending requests without any delays).
* Test scenario: 
  - Start the test and wait for 5 min
  - do full backend rollout and wait for 5 min again

Here are the results:

**round robin**
<img width="1160" alt="Screenshot 2023-09-08 at 9 59 35 AM" src="https://github.com/grpc/proposal/assets/7030896/aa3668a7-d107-46d6-86a6-9adcc772de08">
Nothing surprising here: all servers get equal amount of connections and requests, there are 2 visibly distinct buckets for normal and noisy clients as well as for fast and slow servers. 

**wrr**
<img width="1160" alt="Screenshot 2023-09-08 at 10 04 06 AM" src="https://github.com/grpc/proposal/assets/7030896/aec01b27-d882-4a24-9023-8be4d2b45820">
As expected, wrr was able to fix backend CPU imbalance by adjusting corresponding weights, now slow servers receive less requests than fast servers, but CPU load is well aligned (with ~4% difference)

**random subsetting + wrr (subset size = 10)**
<img width="1160" alt="Screenshot 2023-09-08 at 11 00 18 AM" src="https://github.com/grpc/proposal/assets/7030896/dccbe04b-e575-41d3-bcd6-c9e867dbdfc0">


First of all, the resulting number of connections look quite bad. Second, wrr with default load function doesn't actually do anything to make the resulting CPU load better. WRR only takes into account "cost per request", which is different for slow and fast servers, but is identical for the servers that are in the same group but happen to receive different number of connections. 

To fix this problem I changed the definition of CPULoad parameter and multiplied it by the number of incoming connections. (this is the thing I call "scaled wrr")

**random subsetting + scaled wrr (subset size = 10)**
Here I must admit that just scaled wrr version works very well if we have a single type of clients. If, however, we have both noisy and normal clients it doesn't work well. If a backend have 2 noisy connections and 1 normal ideally it should get a different weight than the server that get 1 noisy and 2 normal connections. To deal with this problem I added an interceptor that writes client type to metadata header and grouped the connections by client_type on the server. Then I returned different ORCA responses for normal and noisy clients (the scaling factor for a particular client now take into account only the number of incoming connections  of the same type) In order to do that I had to fork grpc and slightly modify ORCA producer interface. The results are the following:
<img width="1160" alt="Screenshot 2023-09-08 at 10 33 16 AM" src="https://github.com/grpc/proposal/assets/7030896/1c913bd4-130a-41ef-b470-3856319d53e3">
The results are much better but still not perfect. There are a few clear outliers with lower CPU usage then the rest of the backends. My guess is that those outliers are "lucky" in a sense that they got connections only from normal servers and wrr can't do anything to put more load on them. If this is true - this is a problem of subset size. So in the next test I bumped it to 30. 

**random subsetting + scaled wrr (subset size = 30)**
<img width="1160" alt="Screenshot 2023-09-08 at 11 16 51 AM" src="https://github.com/grpc/proposal/assets/7030896/1c348f89-1dcc-4e42-8d5e-283479472985">
 Now we got perfect results visually indistinguishable from plain wrr. However, figuring out the right subset size for a particular workload in this scenario might be challenging and for some types of workloads we may need realatively large subset sizes, which is not ideal.

**Open questions**
* We didn't explore the option to dynamically rebalance the connections in the subsetting load balancer based on the number of incoming connections per server. This will require making subsetting LB aware of ORCA metrics as well as some changes to ORCA setup. The potential benefit is that it could probably help us to achieve convergence on much smaller subset sizes. But I wold rather not spend time exploring this options if that is something gRPC maintainers will never accept.
* Should we make ORCA capable of returning different metrics depending on the client? If yes, this probably should go to a different gRFC, but the change itself is pretty small. 

**Proposed plan**
Given the shortcomings of the random subsetting + wrr approach we propose to get back to deterministic subsetting implementation. In order to address the major concern about propagating cleint_id to the clients we propose to do the following:
* Get client ID from an environment variable (something like GRPC_SUBSETTING_CLEINT_ID) This guaranties that the ID is stable.
* In multi-cluster k8s environment we will do the following to generate this ID
  - Treat the same clients from different k8s clusters as different set of clients and assign IDs for them independently. 
  - Use admission webhook that watches all pods for a deployment and assigns the next available integer ID as soon as it detects a new pod. Then modify the pod spec and inject the ID as ENV.
  - The shortcomings of the described approach is that the resulting IDs won't start from 0, in some cases the indexes may have gaps (for example if a pod get manually deleted)  and we will artificially split the clients across k8s clusters boundaries. All this will result in a worse connection per backend distribution, but my intuition is that it still will be way better than with random subsetting. We will use "scaled wrr" to correct the resulting connection imbalance, but given that this imbalance will be much smaller than with random subsetting we will be able to use much smaller subsets. And this should work for any type of workload. 
 
@markdroth what do you think?

Reply:
Some quick notes (which I failed to send last week):
 * Even with deterministic subsetting you will get banding. I'd suggest choosing a subset no smaller than 20, so that the bands are 5% of the load
 * Yes, WRR doesn't balance utilization. Except it _does_ balance utilization once your servers start becoming slower due to load. You have to be running "hot" for WRR to be different than RR (unless you start hacking the qps/utilization ratio :wink:). If a 20% utilized server behaves the same as an 80% utilized server, then it doesn't matter to the client which it uses... but it is bothersome to manage the servers
 * Multiplying utilization by number of connections is an interesting approach, and wouldn't suffer from oscillations but doesn't seem like something we should recommend in general. There's probably a decision here whether to consider a balancer that tries to balance based on utilization, before the backends become saturated. I'm not familiar with prior art here, but I expect such a balancer has to take pains to avoid oscillations.

Reply:
@ejona86, all you've said is consistent with my understanding. A few notes:
* Waiting until servers become slower due to load is not an acceptable option for us. From my experiments looks like this happens only at a very high load, which can also result in elevated error rate and latency. Also we have some workloads that are spiky in nature, so app owners have to over-provision to account for spikes. So we should find a solution that results in even load distribution, even when backends are not fully utilized. 

> Multiplying utilization by number of connections is an interesting approach, and wouldn't suffer from oscillations but doesn't seem like something we should recommend in general.

I agree that this is hacky, but do you have any specific use-cases in mind where this approach could cause issues?

In general we are waiting for the direction here. We can try using deterministic subsetting + wrr, or work on a balancer that tries to balance based on utilization, before the backends become saturated (we already partnered with "applied science" team at DataDog to work on that) If you have any better ideas we can also test them out.

Reply:
TL;DR; of our offline subsetting discussion:
* Any form of deterministic subsetting requires coordination between clients, and if we need coordination between clients this should be done via the control-plane. But if control-plane is a requirement for the subsetting we can also run subsetting algorithm in the control-plane, which is the recommended way of doing deterministic subsetting. Doing deterministic subsetting inside a client-side LB policy is not an acceptable option. 
* Random subsetting is ok.
* We should look into how we can improve imbalance generated by random subsetting. Scaling wrr by the number of connections is not ideal because it is not generic enough - instead we should look into algorithms that balance based on load and not on server capacity (like wrr does) Fighting oscillations is the major problem here. 

To address the last point I want to test how [Proportional–integral–derivative controller](https://en.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller) works for balancing the load. It was suggested to me by some folks at gRPC conf and it is actually used in practice by some companies to correct the imbalance generated by random subsetting. It is also specifically designed to prevent oscillations and provide some mathematical guaranties about that. 

If PID controller approach doesn't work or end up being not acceptable either I'll five up on this proposal.

Reply:
We've got some really interesting results while testing [PID controller](https://en.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller) in combination with random subsetting. After we test it with real apps in production, I am going to close this proposal and instead open a new one, which will define 2 new LB policies:
* Random subsetting using rendez-vous hashing.
* PID controller LB policy.

@markdroth @ejona86 does this sound ok to you?

**PID controller test results**
Before I proceed to the results I want to highlight a few important details about my PID controller implementation
* It is based on wrr, it uses signal from PID controller to modify wrr weights. 
* It stores a PID controller per wrr sub-channel, The controller is configured to use mean CPU utilization over all connected upstream servers as target value. 
* I used the following code to convert the signal from PID controller to wrr weight
```
        w.pidController.Update(pid.ControllerInput{
		ReferenceSignal:  meanUtilization,
		ActualSignal:     utilization,
		SamplingInterval: time.Since(w.lastUpdated),
	})
	if w.weightVal == 0 {
		w.weightVal = 1
	}
	mult := 1.0
	signal := w.pidController.State.ControlSignal
	if signal >= 0 {
		mult = 1.0 + signal
	} else {
		mult = -1.0 / (signal - 1.0)
	}
	w.weightVal *= mult
```
If the signal is positive we slowly increase the weight, proportionally to the value of the signal. The opposite happens is the signal is negative
* I used only proportional and derivative part of the PID controller (and ignored the integral pert) Integrals accumulate over time and need to be discharged somehow. There are technics to do that, but the didn't show any improvements in comparison to a simple PD controller. I think that integral part is really important if the target value changes sharply. In our case target value is meanUtilization, which should change only gradually. (We actually applied moving average while reporting CPU load to make sure that utilization metric is smooth, this is very important for PID controller to avoid oscillations. I'll cover this in more details later)
* When using PID controller weight update period must be synchronized with reporting period, otherwise we risk updating the internal state of the controller  with values that are not actually applied. I achieved this by using out-of-band reporting with the same interval as weight update period, we can achieve the same result with in-bound reporting as well.
* The resulting PID controller LB exposes only 2 additional parameters in comparison to wrr LB: ProportionalGain and. DerivativeGain. It is easy to tune then: Increasing ProportionalGain reduces the time to converge but increases the chance of oscillations. Increasing DerivativeGain makes the convergence more smooth and reduces the chance of overshoot and oscillations, but if set too high it can slow down convergence. 
* We can provide some guaranties about non oscillating when using PID controller. The probability of oscillations depends on:
  * Delay between correction action and getting corresponding load results (depends on the frequency of weight updates and load reports)
  * ProportionalGain (the higher the value the higher the probability of oscillations)
  * DerivativeGain (the lower the value the higher the probability of oscillations)
  * How fast utilization changes on the server (can be capped by using moving average either on the server or on the client)
  * How fast meanUtilization changes (can be capped by using moving average either on the server or on the client)
The probability of oscillations only indirectly depends  on QPS and load profile generated by individual clients (via changes to server utilization and meanUtilization) Using moving average will cut the noize in utilization. All other parameters are hardcoded and can be tuned in a generic way for any types of workloads. 

**Test setup**
This time I run 2 different tests:
* Test 1
  * 93 clients, 87 server
  * 2 types of clients (normal and noizy)
  * 2 types of servers (fast and slow)
* Test 2
  * 93 clients, 87 server
  * 2 types of clients (normal and spiky)
    * Spiky clients increase load by sinusoid with 1 min rate
  * 2 types of servers (fast and slow)
  * 3 types of request(light, medium, heavy) with random distribution
The second test is needed to validate that PID controller can adopt to the changes in load quickly enough (As you will see from the graph the time between spikes was ~1 min)

**Test 1: plain wrr with no subsetting**
<img width="1163" alt="Screenshot 2023-09-27 at 9 50 26 PM" src="https://github.com/grpc/proposal/assets/7030896/ea80a66d-73d7-4de1-a8cf-be159e08b92b">
I use this test as a reference point, we can't do bettter than that.

**Test 1: PID controller + subsetting (subset_size = 20)**
<img width="1163" alt="Screenshot 2023-09-27 at 9 51 45 PM" src="https://github.com/grpc/proposal/assets/7030896/3ce0eded-0c5c-4a99-8951-1a9269dc5eb9">
As you can see, the load converges to almost a perfect straight line after ~1 min. 

**Test 2: plain wrr with no subsetting**
<img width="1163" alt="Screenshot 2023-09-27 at 9 53 11 PM" src="https://github.com/grpc/proposal/assets/7030896/f20f1b16-c82c-4e8c-a2c6-80418fcc4254">

**Test 2: PID controller + subsetting (subset_size = 20)**
<img width="1163" alt="Screenshot 2023-09-27 at 9 54 59 PM" src="https://github.com/grpc/proposal/assets/7030896/d69d6d57-1c33-41b4-a230-56d02e595307">
The load distribution during spikes is now as tight as with wrr and no subsetting, but IMO it doesn't look bad at all. The spikes are pretty sort (~ 1 min) and PID controller still can correct the load with no oscillations or overshoot. 

**Test 1: PID controller + subsetting (subset_size = 5)**
<img width="1163" alt="Screenshot 2023-09-27 at 9 57 07 PM" src="https://github.com/grpc/proposal/assets/7030896/dac6b9ae-2972-4cb7-a28c-176b55699787">
I also decided to test what happens we we use very small subset size. As you can see, the resulting load still converges (even though the connection distributions is terribly unbalances in this case. Also, with very small subsets every client have a very limited view of the servers, but given enough subset diversity it still allows the load to converge)

Reply:
I can't fully respond, but I also don't want to leave you hanging without hearing anything back.

Those results look really nice. The PID worked well. Subsetting behaved as expected. You threw a lot of ugly cases at it and it held up. I'm excited.

There's been some separate conversations that just so happen to have happened in parallel which may impact the random subsetting... In a conversation about very low QPS clients we found ourselves talking about dynamic subset sizing. @markdroth was going to figure out how much overlap there was with this, IIRC.

FYI, I'm on vacation Oct 10th through Oct 18th.

Reply:
I finally can report some result here: we tested PID balancer for a few apps in prod environment and learned a few important lessons about how we should tune it. Here is the result of applying PID balancer in one of our production DCs for an app that uses random subsetting.
![Screenshot 2024-04-02 at 3 44 04 PM](https://github.com/grpc/proposal/assets/7030896/6846d995-79b1-4b6e-b2c6-7575af122635)
We want to contribute the code upstream, so we don't have to support a fork of wrr and the rest of the community can reuse it as well. I am going to write 3 gRFCs:
* Add simple client-side random subsetting based on consistent hashing.
* Add moving-average support to server-side ORCA metrics (This is critical for PID to work correctly, otherwise weights won't converge for some spiky workloads)
* Add PID balancer (My implementation reuses 90% of WRR code, so I am going to propose some options how to reuse the code without duplication)

@markdroth @ejona86 does this sound good?

Reply:
Those last two could be in one gRFC, since the PID balancer is to use the new ORCA functionality.